blei-lab/edward
# Feature/criticism

#### Summary:

+ issues: fixes #41; makes progress for #104; relevant to #105

This feature request provides a foundation for the criticism interface. Two primary functions have been added: `ed.evaluate()` and `ed.ppc()`. 

`ed.evaluate()` does point prediction for 15 currently implemented metrics, comparing the model's prediction to the truth. E.g., mean squared error and classification accuracy. `ed.ppc()` does posterior predictive checks, which returns a reference distribution and realized discrepancy for more general distribution-based model assessments. This design is primarily based on @mariru's previous pull request and also the discussion in #104. 

Extensions (to be added as Github issues after merge):
+ __Leave-one-out cross validation.__ Following the explanation in [Vehtari, Gelman, Gabry (2016)](http://arxiv.org/abs/1507.04544), this is a very simple wrapper around `ed.evaluate()`. The only algorithmic change is importance sampling, so that we can use samples from `p(z | y)` so it is as if they came from the leave-one-out posterior `p(z | y_{-i})`.
+ __Graphical checks.__These are PPC's but possibly without a discrepancy, and the return object is a (seaborn?) plot along a certain dimension.
+ __Global sessions.__ It's kind of a pain to plug in all these arguments any time we want to criticize model fit. By default the functions should grab the session globally from the environment, if it exists. To do this, inference will be modified to globally instantiate the session rather than just locally.
+ The current inputs and outputs of `model.predict()` are rather unnatural. In principle, it shouldn't have access to the test output, and only the test input. For this to apply, the Data class must have a notion that distinguishes between output data and input data.
+ Add documentation page explaining criticism.
+ The amount of (optional) methods we use in the model object is getting unwieldy. This pull request uses three more: `predict()`, `sample_prior()`, and `sample_likelihood()`. We should add a documentation page explaining the model object in-depth and these optional methods. For now, I have it written down in [this gist](https://gist.github.com/dustinvtran/1f789de9cce60aed83300739bd9b6579). (The specs here are almost telling us that in the future we may end up writing our own modeling language to automate all of this.)
+ #104 to add training/test phases for PrettyTensor if the user has it installed (training phase is placed inside inference; test phase is placed inside `ed.evaluate()` and `ed.ppc()`)

#### Intended Effect:

See above.

#### How to Verify:

Run `bayesian_linear_regression_test.py` to see `ed.evaluate()` in action. Run `beta_bernoulli_ppc.py` to see `ed.ppc()` in action.

#### Side Effects:

N/A

#### Documentation:

I added comments for any non-trivial lines of code. I think to explain criticism well requires a stand-alone documentation page; this in turn makes the move to a new documentation interface even more important.

#### Reviewer Suggestions:

@mariru
